### ISSUE TITLE
"persistenceMode:true" blocking error "TypeError: localStorage is null"

### ISSUE THREAD

**Issue by G-WAN** [olifolkerd/tabulator#2271]
**Describe the bug**
Tabulator.js does not check the case of a null localStorage (which happens if localStorage is disabled by the Internet browser options - here Firefox, 68.0.1, latest version).

**Tabulator Info**
4.3 (latest)
- copy of your construct object (works with any table...):

 tablexxx = new Tabulator("#ex-table",{
 height:"100%",
 virtualDom:true,
 resizableColumns:true,
layout:"fitDataFill",
 selectable:true,
 pagination:"local", 
 paginationSize:14,
 paginationButtonCount:1,
 persistenceID:"tablex",
//persistenceMode:true,    // "cookie" if "local" (storage) unavailable
 persistenceMode:"cookie", // true fails: "TypeError: localStorage is null"
 persistentLayout:true,
 });

**Comment by aklaver** [comment]
Hmm:
```
window.localStorage
null
mode
"cookie"
m = mode = mode!== true ?  mode : (typeof window.localStorage !== 'undefined' ? "local" : "cookie");
"cookie"
```

**Comment by aklaver** [comment]
Aah I see the problem. The code will set the storage to 'local' if persistenceMode is set to true or 'local' even if localStorage is not available. Later when it persists data it will fail with 'TypeError: localStorage is null'.  In FireFox when localStorage is disabled it is set to null. In Chrome localStorage is not actually disabled, it just cannot be written to by outside sites. I don't know what the other browsers do.  The storage test needs to change to deal with this. Not sure how at the moment.

**Comment by G-WAN** [comment]
Maybe something [like](https://stackoverflow.com/questions/9847580/how-to-detect-safari-chrome-ie-firefox-and-opera-browser):

```

/**
 * Gets the browser name or returns an empty string if unknown. 
 * This function also caches the result to provide for any 
 * future calls this function has.
 *
 * @returns {string}
 */
var browser = function() {
    // Return cached result if avalible, else get result then cache it.
    if (browser.prototype._cachedResult)
        return browser.prototype._cachedResult;

    // Opera 8.0+
    var isOpera = (!!window.opr && !!opr.addons) || !!window.opera || navigator.userAgent.indexOf(' OPR/') >= 0;

    // Firefox 1.0+
    var isFirefox = typeof InstallTrigger !== 'undefined';

    // Safari 3.0+ "[object HTMLElementConstructor]" 
    var isSafari = /constructor/i.test(window.HTMLElement) || (function (p) { return p.toString() === "[object SafariRemoteNotification]"; })(!window['safari'] || safari.pushNotification);

    // Internet Explorer 6-11
    var isIE = /*@cc_on!@*/false || !!document.documentMode;

    // Edge 20+
    var isEdge = !isIE && !!window.StyleMedia;

    // Chrome 1+
    var isChrome = !!window.chrome && !!window.chrome.webstore;

    // Blink engine detection
    var isBlink = (isChrome || isOpera) && !!window.CSS;

    return browser.prototype._cachedResult =
        isOpera ? 'Opera' :
        isFirefox ? 'Firefox' :
        isSafari ? 'Safari' :
        isChrome ? 'Chrome' :
        isIE ? 'IE' :
        isEdge ? 'Edge' :
        isBlink ? 'Blink' :
        "Don't know";
};

console.log(browser());
```

**Comment by aklaver** [comment]
@G-WAN thanks for the example. I leaning more to something like this:
https://stackoverflow.com/questions/16427636/check-if-localstorage-is-available
```
function lsTest(){
    var test = 'test';
    try {
        localStorage.setItem(test, test);
        localStorage.removeItem(test);
        return true;
    } catch(e) {
        return false;
    }
}
```
The key is whether localStorage is available for use or not and the above checks for that.

**Comment by aklaver** [comment]
PR:
https://github.com/olifolkerd/tabulator/pull/2280

**Comment by G-WAN** [comment]
Excellent! When do you think a new release will be made available? (Firefox issue about moveable columns from one table to another is lagging since 2018).

**Comment by aklaver** [comment]
I do not control  the release schedule. For information from the person who is see:
https://github.com/olifolkerd/tabulator/issues/2213

The Firefox issue is tied to this issue I filed with Mozilla:
https://bugzilla.mozilla.org/show_bug.cgi?id=1529753
As you can see that has stalled.

**Comment by G-WAN** [comment]
I was referring to the Tabulator bug regarding Persistence. When do you think your patch will be published?

**Comment by aklaver** [comment]
First the PR has to be accepted and then released. That is not under my control. If you need the change you could cut and paste the code from the PR link above into you local copy of tabulator.js. The code is for the persistence.js module source so the line numbers will not be the same as in tabulator.js. In there it is about line 20287. To be on the safe side I would do this on a copy of tabulator.js and then substitute that in.

**Comment by olifolkerd** [comment]
Pull requests for bug fixes will be merged in this weekend for the 4.4 release on Sunday 😁

**Comment by G-WAN** [comment]
Thank you Aklaver and Oli for the informative and pertinent answers.
